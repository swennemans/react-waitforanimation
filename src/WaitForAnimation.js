/* @flow */
import React from 'react';
const vendors = ['transitionend', 'OTransitionEnd', 'webkitTransitionEnd'];

class WaitForAnimation extends React.Component {
    props: {
      componentDidAnimate: () => void,
      children?: any;  // FIXME:, see facebook/flow#1355
    };
    state: {
      isResetted: boolean,
      isEnded: boolean,
    };
    transitionHasEnded: () => void;
    resetAnimationState: () => void;

    constructor(props) {
      super(props);
      this.state = { 
        isEnded: false,
        isResetted: false,
      };
      this.transitionHasEnded = this.transitionHasEnded.bind(this);
      this.resetAnimationState = this.resetAnimationState.bind(this);
    }

    componentDidMount(): void {
      if (this._containerRef !== null) {
        vendors.forEach(tEvent => {
          this._containerRef.addEventListener(tEvent, () => {
            this.transitionHasEnded();
          }, false);
        })
      }
    }

    componentWillUnmount(): void {
      vendors.forEach(tEvent => {
        this._containerRef.removeEventListener(tEvent);
      })
    }
    /**
     * We're only interested in passing animationDidFinish = true
     * to child component when we're coming from !isEnded -> isEnded. 
     * Not when isEnded is already true. This because the transition events API
     * is very limited. For example the: 'transistionstart'' event only works on IE (of all browsers :D)
     * Also, this will allow you to keep your fetched data in animated component without refetching
     * it on every animation. If you dont want this behaviour use this.resetAnimationState first. 
     */
    transitionHasEnded(): void {
      const { isEnded, isResetted } = this.state;
      if (isResetted) {
        this.setState({ isResetted: false });
      } else if (!isEnded) {
        this.setState({ isEnded: true });
        this.props.componentDidAnimate();
      } else {
        this.setState({ isEnded: false });
      }
    }

    /**
     * Set isResetted flag to true. Because we **only** we have
     * a transistionEnd event and it doesn't care if it's our opening
     * or closing animation. We can block a state update in
     * this.transitionHasEnded().
     */
    resetAnimationState(): void {
      if (!this.state.isResetted) {
        this.setState({ isRestted: true });
      }
    }

    render(): React.Element {
      return (
        <div ref={ref => this._containerRef = ref}>
          {this.props.children(this.resetAnimationState)}
        </div>
      )
    }
}

export default WaitForAnimation;
