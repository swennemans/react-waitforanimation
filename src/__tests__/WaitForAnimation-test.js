import WaitForAnimation from '../WaitForAnimation.js';

// see: https://github.com/facebook/jest/issues/1353
// could possibly be removed because removed enzyme. 
beforeEach(() => jest.resetModules());

 describe('EventListener.', () => { 
  const React = require('react');
  const TestUtils = require('react-addons-test-utils');
  const sinon = require('sinon');
  
  sinon.spy(WaitForAnimation.prototype, 'transitionHasEnded');
  
  const mockFn = jest.fn();
  
  test('isEnded should start as false', () => {
    //initiate here because otherwise state has already changed. because of async??
    const comp = TestUtils.renderIntoDocument(<WaitForAnimation componentDidAnimate={mockFn}>
    {(reset) => <div></div>}
    </WaitForAnimation>)
    expect(comp.state.isEnded).toBe(false);
  })

  const comp = TestUtils.renderIntoDocument(<WaitForAnimation componentDidAnimate={mockFn}>
  {(reset) => (
      <div></div>
    )}
  </WaitForAnimation>)

  //** setup + dispatch event *//
  const transistionEvent = new Event('transitionend', { bubbles: false });
  comp._containerRef.dispatchEvent(transistionEvent)

  test('transitionHasEnded should be called', () => {
    expect(WaitForAnimation.prototype.transitionHasEnded.calledOnce).toBe(true);
  })

  test('isEnded should be set to true', () => {
    expect(comp.state.isEnded).toBe(true);
  })

  test('componentDidAnimate prop should be called', () => {
    expect(mockFn).toHaveBeenCalled();
  })

})

test('Component should render', () => {
  const React = require('react');
  const renderer = require('react-test-renderer');

  const mockFn = jest.fn();
  const component = renderer.create(
    <WaitForAnimation componentDidAnimate={mockFn}>
    {(reset) => <div></div> }
    </WaitForAnimation>
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});



