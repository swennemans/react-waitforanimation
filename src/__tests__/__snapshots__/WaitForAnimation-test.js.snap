exports[`test Component should render 1`] = `
<div>
  <div />
</div>
`;
